### ISSUE TITLE
windows_printer - NoMethodError - undefined method `ipv4_address' for nil:NilClass

### ISSUE THREAD

**Issue by pauby** [chef-cookbooks/windows#521]
### Cookbook version
3.4.3

### Chef-client version
13.6.4

### Platform Details
Running from Windows 10 workstation. Cookbook is being applied to a **different** Windows 10 workstation (running under Hyper-V)

### Scenario:

Using windows_printer in recipe which gives the error (using kitchen converge):
```
===============================
Error executing action `create` on resource 'windows_printer_port[192.168.10.30]'
===============================

NoMethodError
-------------
undefined method `ipv4_address' for nil:NilClass
```

### Steps to Reproduce:
The recipes being used is an almost copy of your example:

```
windows_printer 'HP LaserJet P2015 Printer' do
  driver_name 'HP LaserJet P2015 PCL6 Class Driver'
  ipv4_address '192.168.10.30'
  location 'Office'
end
```

Running `kitchen converge` gives:

```
-----> Starting Kitchen (v1.17.0)
-----> Converging <default-win10-x64>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 6.3.1...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Chef Omnibus installation detected (install only if missing)
       Transferring files to <default-win10-x64>
       Starting Chef Client, version 13.6.4
       resolving cookbooks for run list: ["base::default"]
       Synchronizing Cookbooks:
         - base (0.1.0)
         - windows (3.4.3)
         - ohai (5.2.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: base::printing
         * windows_printer[HP LaserJet P2015 Printer] action create
           - Create windows_printer[HP LaserJet P2015 Printer]
           * windows_printer_port[192.168.10.30] action create

             ================================================================================
             Error executing action `create` on resource 'windows_printer_port[192.168.10.30]'
             ================================================================================

             NoMethodError
             -------------
             undefined method `ipv4_address' for nil:NilClass

             Resource Declaration:
             ---------------------
             # In C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/printer.rb

       72:     windows_printer_port new_resource.ipv4_address do
       73:     end
       74:

             Compiled Resource:
             ------------------
             # Declared in C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/printer.rb:72:in `create_printer'

             windows_printer_port("192.168.10.30") do
        action [:create]
        default_guard_interpreter :default
        declared_type :windows_printer_port
        cookbook_name "base"
        ipv4_address "192.168.10.30"
             end

             System Info:
             ------------
             chef_version=13.6.4
             platform=windows
             platform_version=10.0.14393
             ruby=ruby 2.4.2p198 (2017-09-14 revision 59899) [x64-mingw32]
             program_name=C:/opscode/chef/bin/chef-client
             executable=C:/opscode/chef/bin/chef-client


           ================================================================================
           Error executing action `create` on resource 'windows_printer[HP LaserJet P2015 Printer]'
           ================================================================================

           NoMethodError
           -------------
           windows_printer_port[192.168.10.30] (C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/printer.rb line 72) had an error: NoMethodError: undefined method `ipv4_address' for nil:NilClass

           Resource Declaration:
           ---------------------
           # In C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/base/recipes/printing.rb

             6: windows_printer 'HP LaserJet P2015 Printer' do
             7:   driver_name 'HP LaserJet P2015 PCL6 Class Driver'
             8:   ipv4_address '192.168.10.30'
             9:   location 'Office'
            10: end

           Compiled Resource:
           ------------------
           # Declared in C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/base/recipes/printing.rb:6:in `from_file'

           windows_printer("HP LaserJet P2015 Printer") do
             action [:create]
             default_guard_interpreter :default
             declared_type :windows_printer
             cookbook_name "base"
             recipe_name "printing"
             driver_name "HP LaserJet P2015 PCL6 Class Driver"
             ipv4_address "192.168.10.30"
             device_id "HP LaserJet P2015 Printer"
           end

           System Info:
           ------------
           chef_version=13.6.4
           platform=windows
           platform_version=10.0.14393
           ruby=ruby 2.4.2p198 (2017-09-14 revision 59899) [x64-mingw32]
           program_name=C:/opscode/chef/bin/chef-client
           executable=C:/opscode/chef/bin/chef-client


       Running handlers:
       [2018-01-06T22:43:02+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2018-01-06T22:43:02+00:00] ERROR: Exception handlers complete
       Chef Client failed. 0 resources updated in 04 seconds
       [2018-01-06T22:43:02+00:00] FATAL: Stacktrace dumped to C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/chef-stacktrace.out
       [2018-01-06T22:43:02+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
       [2018-01-06T22:43:02+00:00] FATAL: NoMethodError: windows_printer[HP LaserJet P2015 Printer] (base::printing line 6) had an error: NoMethodError: windows_printer_port[192.168.10.30] (C:/Users/ADMINI~1/AppData/Local/Temp/kitchen/cache/cookbooks/windows/resources/printer.rb line 72) had an error: NoMethodError: undefined method `ipv4_address' for nil:NilClass
```

### Expected Result:
Printer to be created or at the very least no error.

### Actual Result:
See error.

DIdn't want to provide the stacktrace right now in case it's something straightforward. Just to make sure I had everything setup in Chef I used `windows_shortcut` example from the README (to create a short to Notepad) which worked fine so I know everything is as it should on my build.

**Comment by jmery** [comment]
@pauby I hit this issue and have the fix.  I'll have a PR inbound shortly.

TL;DR The issue is actually with the `windows_printer_port` resource here: https://github.com/chef-cookbooks/windows/blob/96652ca456d87c37937ca8a81b5c725ff603adcc/resources/printer_port.rb#L45-L46.  Those lines needs a little help to better handle the optional `port_name` property.
